File: sdk/node/src/fs.ts
import type { Client } from './client.js'

export type FsEntryKind = 'file' | 'directory' | 'symlink' | 'other'

export interface FsEntry {
  path: string
  kind: FsEntryKind
  size: number
  mode: number
  modified: Date | null
}

export interface FsMetadata {
  kind: FsEntryKind
  size: number
  mode: number
  readonly: boolean
  modified: Date | null
  created: Date | null
}

function parseTime(v: unknown): Date | null {
  if (v == null || v === '') return null
  if (typeof v === 'string') {
    const d = new Date(v)
    return Number.isNaN(d.getTime()) ? null : d
  }
  return null
}

function queryPath(sandboxId: string, op: string, path: string): string {
  const q = new URLSearchParams({ path })
  return `/v1/sandboxes/${encodeURIComponent(sandboxId)}/fs/${op}?${q}`
}

/**
 * Streaming reader for a guest file. Implements AsyncIterable and AsyncDisposable.
 */
export class FsReadStream implements AsyncIterable<Uint8Array>, AsyncDisposable {
  private readonly reader: ReadableStreamDefaultReader<Uint8Array>
  private closed = false

  /** @internal */
  constructor(body: ReadableStream<Uint8Array>) {
    this.reader = body.getReader()
  }

  /** Receive the next chunk, or null at EOF. */
  async recv(): Promise<Uint8Array | null> {
    if (this.closed) return null
    const { done, value } = await this.reader.read()
    if (done) {
      this.closed = true
      return null
    }
    return value ?? new Uint8Array()
  }

  /** Drain the stream into a single buffer. */
  async collect(): Promise<Uint8Array> {
    const parts: Uint8Array[] = []
    let total = 0
    for (;;) {
      const chunk = await this.recv()
      if (chunk == null) break
      parts.push(chunk)
      total += chunk.byteLength
    }
    const out = new Uint8Array(total)
    let offset = 0
    for (const p of parts) {
      out.set(p, offset)
      offset += p.byteLength
    }
    return out
  }

  async *[Symbol.asyncIterator](): AsyncIterator<Uint8Array> {
    for (;;) {
      const chunk = await this.recv()
      if (chunk == null) return
      yield chunk
    }
  }

  async [Symbol.asyncDispose](): Promise<void> {
    if (this.closed) return
    this.closed = true
    try {
      await this.reader.cancel()
    } catch {
      // ignore
    }
  }
}

/**
 * Streaming writer for a guest file. Implements AsyncDisposable.
 */
export class FsWriteSink implements AsyncDisposable {
  private readonly writer: WritableStreamDefaultWriter<Uint8Array>
  private readonly done: Promise<void>
  private closed = false

  /** @internal */
  constructor(writer: WritableStreamDefaultWriter<Uint8Array>, done: Promise<void>) {
    this.writer = writer
    this.done = done
  }

  /** Append a chunk. Strings are encoded as UTF-8. */
  async write(data: Uint8Array | string): Promise<void> {
    if (this.closed) throw new Error('write stream is closed')
    const chunk = typeof data === 'string' ? new TextEncoder().encode(data) : data
    await this.writer.write(chunk)
  }

  /** Flush and close. Idempotent. */
  async close(): Promise<void> {
    if (this.closed) return
    this.closed = true
    await this.writer.close()
    await this.done
  }

  async [Symbol.asyncDispose](): Promise<void> {
    await this.close()
  }
}

/** In-sandbox filesystem operations for a single sandbox. */
export class SandboxFs {
  private readonly client: Client
  private readonly sandboxId: () => string

  /** @internal */
  constructor(client: Client, sandboxId: () => string) {
    this.client = client
    this.sandboxId = sandboxId
  }

  /** Read the entire contents of a file as raw bytes. */
  async read(path: string): Promise<Uint8Array> {
    const stream = await this.readStream(path)
    try {
      return await stream.collect()
    } finally {
      await stream[Symbol.asyncDispose]()
    }
  }

  /** Read the entire contents of a file and decode as UTF-8. */
  async readToString(path: string): Promise<string> {
    const bytes = await this.read(path)
    return new TextDecoder().decode(bytes)
  }

  /** Write content to a file (create or overwrite). Parents must exist. */
  async write(path: string, data: Uint8Array | string): Promise<void> {
    const body = typeof data === 'string' ? new TextEncoder().encode(data) : data
    await this.client.fsPutContent(this.sandboxId(), path, body)
  }

  /** Open a streaming reader for a file. */
  async readStream(path: string): Promise<FsReadStream> {
    const body = await this.client.fsGetContentStream(this.sandboxId(), path)
    return new FsReadStream(body)
  }

  /**
   * Open a streaming writer for a file.
   * Prefer `await using sink = await fs.writeStream(path)`.
   */
  async writeStream(path: string): Promise<FsWriteSink> {
    return this.client.fsPutContentStream(this.sandboxId(), path)
  }

  /** List directory entries. */
  async list(path: string): Promise<FsEntry[]> {
    const out = (await this.client.fsJSON(
      'GET',
      queryPath(this.sandboxId(), 'list', path),
    )) as { entries?: Array<Record<string, unknown>> }
    return (out?.entries ?? []).map((e) => ({
      path: String(e.path ?? ''),
      kind: (e.kind as FsEntryKind) ?? 'other',
      size: Number(e.size ?? 0),
      mode: Number(e.mode ?? 0),
      modified: parseTime(e.modified),
    }))
  }

  /** Create a directory. Parents must exist. */
  async mkdir(path: string): Promise<void> {
    await this.client.fsJSON('POST', `/v1/sandboxes/${encodeURIComponent(this.sandboxId())}/fs/mkdir`, {
      path,
    })
  }

  /** Remove an empty directory. */
  async removeDir(path: string): Promise<void> {
    await this.client.fsJSON(
      'POST',
      `/v1/sandboxes/${encodeURIComponent(this.sandboxId())}/fs/remove-dir`,
      { path },
    )
  }

  /** Remove a file. */
  async remove(path: string): Promise<void> {
    await this.client.fsJSON('POST', `/v1/sandboxes/${encodeURIComponent(this.sandboxId())}/fs/remove`, {
      path,
    })
  }

  /** Get file or directory metadata. */
  async stat(path: string): Promise<FsMetadata> {
    const e = (await this.client.fsJSON(
      'GET',
      queryPath(this.sandboxId(), 'stat', path),
    )) as Record<string, unknown>
    return {
      kind: (e.kind as FsEntryKind) ?? 'other',
      size: Number(e.size ?? 0),
      mode: Number(e.mode ?? 0),
      readonly: Boolean(e.readonly),
      modified: parseTime(e.modified),
      created: parseTime(e.created),
    }
  }

  /** Check whether a path exists. */
  async exists(path: string): Promise<boolean> {
    const out = (await this.client.fsJSON(
      'GET',
      queryPath(this.sandboxId(), 'exists', path),
    )) as { exists?: boolean }
    return Boolean(out?.exists)
  }

  /** Copy a file within the sandbox. */
  async copy(from: string, to: string): Promise<void> {
    await this.client.fsJSON('POST', `/v1/sandboxes/${encodeURIComponent(this.sandboxId())}/fs/copy`, {
      from,
      to,
    })
  }

  /** Rename or move a path within the sandbox. */
  async rename(from: string, to: string): Promise<void> {
    await this.client.fsJSON('POST', `/v1/sandboxes/${encodeURIComponent(this.sandboxId())}/fs/rename`, {
      from,
      to,
    })
  }
}

File: sdk/go/fs.go
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

// FsEntryKind matches gateway kind strings.
type FsEntryKind string

const (
	FsKindFile      FsEntryKind = "file"
	FsKindDirectory FsEntryKind = "directory"
	FsKindSymlink   FsEntryKind = "symlink"
	FsKindOther     FsEntryKind = "other"
)

// FsEntry is a directory listing entry.
type FsEntry struct {
	Path     string
	Kind     FsEntryKind
	Size     int64
	Mode     uint32
	Modified *time.Time
}

// FsMetadata is detailed file metadata.
type FsMetadata struct {
	Kind     FsEntryKind
	Size     int64
	Mode     uint32
	Readonly bool
	Modified *time.Time
	Created  *time.Time
}

// SandboxFS exposes in-sandbox filesystem operations for a Sandbox.
type SandboxFS struct {
	client    *Client
	sandboxID string
}

// FS returns the filesystem handle for this sandbox.
func (s *Sandbox) FS() *SandboxFS {
	return &SandboxFS{client: s.client, sandboxID: s.data.ID}
}

func (f *SandboxFS) contentURL(path string) string {
	q := url.Values{"path": {path}}
	return f.client.base + "/v1/sandboxes/" + url.PathEscape(f.sandboxID) + "/fs/content?" + q.Encode()
}

func (f *SandboxFS) opURL(op, path string) string {
	q := url.Values{"path": {path}}
	return "/v1/sandboxes/" + url.PathEscape(f.sandboxID) + "/fs/" + op + "?" + q.Encode()
}

// Read reads the entire file into memory.
func (f *SandboxFS) Read(ctx context.Context, path string) ([]byte, error) {
	rc, err := f.ReadStream(ctx, path)
	if err != nil {
		return nil, err
	}
	defer rc.Close()
	return io.ReadAll(rc)
}

// ReadToString reads the entire file and decodes it as UTF-8.
func (f *SandboxFS) ReadToString(ctx context.Context, path string) (string, error) {
	b, err := f.Read(ctx, path)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// Write writes data to a file (create or overwrite). Parents must exist.
func (f *SandboxFS) Write(ctx context.Context, path string, data []byte) error {
	return f.WriteStream(ctx, path, bytes.NewReader(data))
}

// ReadStream opens a streaming reader for a guest file.
func (f *SandboxFS) ReadStream(ctx context.Context, path string) (io.ReadCloser, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, f.contentURL(path), nil)
	if err != nil {
		return nil, err
	}
	f.client.authHeaders(req.Header)
	req.Header.Set("Accept", "application/octet-stream")
	hc := f.client.http
	// Streaming reads should not inherit a short client timeout.
	if hc.Timeout > 0 {
		clone := *hc
		clone.Timeout = 0
		hc = &clone
	}
	resp, err := hc.Do(req)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		defer resp.Body.Close()
		data, _ := io.ReadAll(resp.Body)
		return nil, parseAPIError(resp.StatusCode, data)
	}
	return resp.Body, nil
}

// WriteStream streams r into a guest file (create or overwrite).
func (f *SandboxFS) WriteStream(ctx context.Context, path string, r io.Reader) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodPut, f.contentURL(path), r)
	if err != nil {
		return err
	}
	f.client.authHeaders(req.Header)
	req.Header.Set("Content-Type", "application/octet-stream")
	hc := f.client.http
	if hc.Timeout > 0 {
		clone := *hc
		clone.Timeout = 0
		hc = &clone
	}
	resp, err := hc.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		data, _ := io.ReadAll(resp.Body)
		return parseAPIError(resp.StatusCode, data)
	}
	return nil
}

// Stat returns metadata for a path.
func (f *SandboxFS) Stat(ctx context.Context, path string) (*FsMetadata, error) {
	var raw map[string]any
	if err := f.client.doJSON(ctx, http.MethodGet, f.opURL("stat", path), nil, &raw); err != nil {
		return nil, err
	}
	return decodeFsMetadata(raw), nil
}

// List lists directory entries.
func (f *SandboxFS) List(ctx context.Context, path string) ([]FsEntry, error) {
	var out struct {
		Entries []map[string]any `json:"entries"`
	}
	if err := f.client.doJSON(ctx, http.MethodGet, f.opURL("list", path), nil, &out); err != nil {
		return nil, err
	}
	entries := make([]FsEntry, 0, len(out.Entries))
	for _, e := range out.Entries {
		entries = append(entries, decodeFsEntry(e))
	}
	return entries, nil
}

// Exists reports whether a path exists.
func (f *SandboxFS) Exists(ctx context.Context, path string) (bool, error) {
	var out struct {
		Exists bool `json:"exists"`
	}
	if err := f.client.doJSON(ctx, http.MethodGet, f.opURL("exists", path), nil, &out); err != nil {
		return false, err
	}
	return out.Exists, nil
}

// Mkdir creates a directory. Parents must exist.
func (f *SandboxFS) Mkdir(ctx context.Context, path string) error {
	return f.client.doJSON(ctx, http.MethodPost,
		"/v1/sandboxes/"+url.PathEscape(f.sandboxID)+"/fs/mkdir",
		map[string]string{"path": path}, nil)
}

// Remove removes a file.
func (f *SandboxFS) Remove(ctx context.Context, path string) error {
	return f.client.doJSON(ctx, http.MethodPost,
		"/v1/sandboxes/"+url.PathEscape(f.sandboxID)+"/fs/remove",
		map[string]string{"path": path}, nil)
}

// RemoveDir removes an empty directory.
func (f *SandboxFS) RemoveDir(ctx context.Context, path string) error {
	return f.client.doJSON(ctx, http.MethodPost,
		"/v1/sandboxes/"+url.PathEscape(f.sandboxID)+"/fs/remove-dir",
		map[string]string{"path": path}, nil)
}

// Copy copies a file within the sandbox.
func (f *SandboxFS) Copy(ctx context.Context, from, to string) error {
	return f.client.doJSON(ctx, http.MethodPost,
		"/v1/sandboxes/"+url.PathEscape(f.sandboxID)+"/fs/copy",
		map[string]string{"from": from, "to": to}, nil)
}

// Rename renames or moves a path within the sandbox.
func (f *SandboxFS) Rename(ctx context.Context, from, to string) error {
	return f.client.doJSON(ctx, http.MethodPost,
		"/v1/sandboxes/"+url.PathEscape(f.sandboxID)+"/fs/rename",
		map[string]string{"from": from, "to": to}, nil)
}

func decodeFsTime(v any) *time.Time {
	if v == nil {
		return nil
	}
	s, ok := v.(string)
	if !ok || s == "" {
		return nil
	}
	t, err := time.Parse(time.RFC3339Nano, s)
	if err != nil {
		t, err = time.Parse(time.RFC3339, s)
		if err != nil {
			return nil
		}
	}
	u := t.UTC()
	return &u
}

func decodeFsMetadata(raw map[string]any) *FsMetadata {
	m := &FsMetadata{
		Kind:     FsEntryKind(fmt.Sprint(raw["kind"])),
		Readonly: raw["readonly"] == true,
		Modified: decodeFsTime(raw["modified"]),
		Created:  decodeFsTime(raw["created"]),
	}
	m.Size = jsonNumberInt64(raw["size"])
	m.Mode = uint32(jsonNumberInt64(raw["mode"]))
	return m
}

func decodeFsEntry(raw map[string]any) FsEntry {
	return FsEntry{
		Path:     fmt.Sprint(raw["path"]),
		Kind:     FsEntryKind(fmt.Sprint(raw["kind"])),
		Size:     jsonNumberInt64(raw["size"]),
		Mode:     uint32(jsonNumberInt64(raw["mode"])),
		Modified: decodeFsTime(raw["modified"]),
	}
}

func jsonNumberInt64(v any) int64 {
	switch n := v.(type) {
	case float64:
		return int64(n)
	case json.Number:
		i, _ := n.Int64()
		return i
	case string:
		i, _ := strconv.ParseInt(n, 10, 64)
		return i
	default:
		return 0
	}
}
